### ISSUE TITLE
Cannot use untyped informers along with typed API

### ISSUE THREAD

**Issue by andreaTP** [fabric8io/kubernetes-client#3785]
### Describe the bug

Using informers for `GenericKubernetesResources` along with accesses to the typed API result in deserialization errors (of the `GenericKubernetesResource`) showing that the typed `Deserializer` has been invoked.

### Fabric8 Kubernetes Client version

next (development version)

### Steps to reproduce

(with access to a cluster e.g. minikube):

1. Clone [https://github.com/andreaTP/repro_GenericKubernetesResource_deser](https://github.com/andreaTP/repro_GenericKubernetesResource_deser)
2. `mvn clean compile`
3. `kubectl apply -f target/classes/META-INF/fabric8/examples.org.example-v1.yml`
4. `mvn clean compile exec:java`

### Expected behavior

The informer should deserialize the `GenericKubernetesResource` without trying to use the "typed" deserializer.

### Runtime

Kubernetes (vanilla), minikube, other (please specify in additional context)

### Kubernetes API Server version

next (development version)

### Environment

macOS

### Fabric8 Kubernetes Client Logs

```shell
2022-01-25 14:42:22,600 ERROR [AbstractWatchManager] - Unhandled exception encountered in watcher event handler
io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:103)
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:97)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:319)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:240)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:226)
        at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.readWatchEvent(AbstractWatchManager.java:237)
        at io.fabric8.kubernetes.client.dsl.internal.AbstractWatchManager.onMessage(AbstractWatchManager.java:262)
        at io.fabric8.kubernetes.client.dsl.internal.WatcherWebSocketListener.onMessage(WatcherWebSocketListener.java:68)
        at io.fabric8.kubernetes.client.okhttp.OkHttpWebSocketImpl$BuilderImpl$1.onMessage(OkHttpWebSocketImpl.java:96)
        at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:323)
        at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:274)
        at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:214)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `int` from String "hello": not a valid `int` value
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: io.fabric8.kubernetes.api.model.WatchEvent["object"]->org.test.v1alpha1.crd.Example["spec"]->org.test.v1alpha1.crd.ExampleSpec["foo"])
        at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)
        at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1991)
        at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:1219)
        at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseIntPrimitive(StdDeserializer.java:768)
        at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseIntPrimitive(StdDeserializer.java:747)
        at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:529)
        at com.fasterxml.jackson.databind.deser.std.NumberDeserializers$IntegerDeserializer.deserialize(NumberDeserializers.java:506)
        at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138)
        at io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:313)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
        at io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:313)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
        at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4650)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2831)
        at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:3295)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.fromObjectNode(KubernetesDeserializer.java:99)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:63)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:45)
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
        at io.fabric8.kubernetes.client.utils.serialization.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:131)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:313)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:176)
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3674)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:317)
        ... 16 common frames omitted
```

### Additional context

Found this issue while trying to support multiple CRD versions in the JO SDK.
It's extremely nasty to debug since removing the access to the "typed" API everything works as expected.

cc. @shawkins @metacosm @csviri

**Comment by shawkins** [comment]
> It's extremely nasty to debug since removing the access to the "typed" API everything works as expected.

In this scenario there are multiple crd versions, but only one typed model, so parsing will fail when the KubernetesDeserializer chooses to deserialize based upon name alone.

You either have to avoid the typed api, or introduce different typed models - one for each crd version.

The informer works around this currently by converting the watchevent item to generic - after it has been parsed.  That only works if it can be correctly parsed by the preregistered type. 

Possible changes
- KubernetesDeserializer should only match on just name when the version is not specified
- If no version is specified on the resource the KubernetesDeserializer should sort the name matching based upon version like crds do.
- Remove the use of KubernetesDeserializer.registerKind - allow the context to specify the expected type, or simply parse to generic and allow the caller to convert to something more specific.

**Comment by andreaTP** [comment]
@shawkins thanks for the prompt answer!

> You either have to avoid the typed api, or introduce different typed models - one for each crd version.

Is it doable in a programmatic way? e.g. when the models are generic.

> simply parse to generic and allow the caller to convert to something more specific

I like this approach!

**Comment by andreaTP** [comment]
Adding more context:
- registering all the versions of the CRD result in always falling back to `v1alpha1` (which is not the version returned from the API)

**Comment by shawkins** [comment]
It doesn't seem possible in general to bring a typed api into this scenario as it won't be handled as expected.

If you enable trace you see:

```
POST https://192.168.39.42:8443/apis/org.example/v2alpha1/namespaces/default/examples -- to create the v2alpha1 entry
```

If you run again, the list operation from the watch or the typed api call, will issue a get:

```
https://192.168.39.42:8443/apis/org.example/v1alpha1/namespaces/default/examples 
```

That response contains:

```
INFO: {"apiVersion":"org.example/v1alpha1","items":[{"apiVersion":"org.example/v1alpha1","kind":"Example","metadata":{"creationTimestamp":"2022-01-25T19:03:26Z","generation":1,"managedFields":[{"apiVersion":"org.example/v2alpha1","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{".":{},"f:bar":{},"f:baz":{},"f:foo":{}}},"manager":"okhttp","operation":"Update","time":"2022-01-25T19:03:26Z"}],"name":"test-example-v2alpha1","namespace":"default","resourceVersion":"761207","uid":"ca50ae40-4b95-4157-b61c-4ca4ed001fce"},"spec":{"bar":123456,"foo":"hello"}}],"kind":"ExampleList","metadata":{"continue":"","resourceVersion":"761282"}}
```

Which is reportedly v1alpha1, but the content is invalid for v1alpha1.  This seems expected when you request the non-stored version and there is no conversion strategy.

**Comment by shawkins** [comment]
I need to correct my initial thoughts:

> KubernetesDeserializer should only match on just name when the version is not specified

After running through the example, that is happening as expected.

> If no version is specified on the resource the KubernetesDeserializer should sort the name matching based upon version like crds do.

It doesn't look like that is a generally supported case.  If the version is missing for create, kubectl will throw an exception.

**Comment by shawkins** [comment]
To sum up - if you request the stored version, or request a version that is appropriately converted to its representation, then the current fabric8 logic will work.  

Do you think that you will need to support the case where you request the non-stored version such that it won't be appropriately converted?

> Remove the use of KubernetesDeserializer.registerKind

It would still be great to eliminate eventually or at least enhance the contextual awareness.

**Comment by shawkins** [comment]
Resolved with #3786
